docs/design-documents/20221027-builtin-config-validation.md

```diff
@@ -0,0 +1,144 @@
+      // Parameter is a number.
+      TYPE_NUMBER = 2;
+      // Parameter is a boolean.
+      TYPE_BOOL = 3;
+      // Parameter is a file.
+      TYPE_FILE = 4;
+      // Parameter is a duration.
+      TYPE_DURATION = 5;
+    }
+
+    string description = 1;
+    string default = 2;
+    Type type = 3;
+    repeated Validation validations = 4;
+  }
+
+  string name = 1;
+  string summary = 2;
+  string description = 3;
+  string version = 4;
+  string author = 5;
+  map<string, Parameter> destination_params = 6;
+  map<string, Parameter> source_params = 7;
+}
+```
+
+For now, this is only used when listing the plugins in Conduit. All these validation options are not yet implemented and
+are not exposed by the SDK.
+
+## Scope
+
+Implementing the validation options provided by the proto file. So, giving the developer the option to specify
+validations for each parameter and Conduit will make sure to run the validations.
+
+Validating the type of the parameter, we have 5 types supported in the proto design {string, number, bool, file, and duration}
+
+Providing a utility function to generate the config map for the `Parameters` function from a config struct. This is not 
+mandatory for the scope of this feature, but it would be a nice to have and would make the developing experience for connectors easier.
+
+## Questions:
+
+**Q**: Should we implement the validations from the SDK or Conduit side?
+
+**A**: If we implement it from Conduit side, this means we can add more validations in the future without changing the SDK.
+However, implementing it from the SDK side means that all the validations will happen on the connector side, using the 
+Config function, and the developer will still have the ability to add custom validations for some specific cases that 
+the SDK wouldn't cover. So we decided on implementing validating from the SDK side.
+
+**Q**: Should generating the Configurations from a Go struct be in the scope of this feature?
+
+**A**: We decided to add the generation as a subtask for now, and see how much time left we have for it. We agreed that this
+would be super helpful for developers and easier to use, and will continue adding to the feature in the future.
+
+## Implementation
+
+Implementing this feature consists of three main steps:
+
+1. Give developers the ability to specify validations needed for each parameter. (manually)
+   Params should look something like:
+
+```go
+SourceParams: []sdk.Parameter{
+    {
+      Name: "param",
+      Type: sdk.ParameterTypeNumber,
+      Validations: []sdk.Validation{
+        sdk.Validation{
+          Type: sdk.ValidationRequired,
+        },
+        sdk.Validation{
+          Type: sdk.ValidationGreaterThan,
+          Value: 0,
+        },
+      }
+    }
+  }
+```
+
+2. Provide a function that takes the parameters' validations and validates them in the config function.
+3. Generate connector configurations from a Go struct, which will give the ability to generate the connector's 
```

Comment:
Do we plan to work on parsing? The original issue mentions validation only, but I see some comments mentioning parsing too. I think it would be good to have an SDK utility which would take a map, validate it (using the specs above or the tags here), and then validate, and if valid, parse it.

Reply:
Currently step 2 provides a function for validation, we might want to have another step that provides a function for parsing into a struct. Arguably we could also provide a utility method that does both at the same time.

One could also argue that providing the validation utility is enough because the user can already use something like [`mitchellh/mapstructure`](https://github.com/mitchellh/mapstructure) for parsing the config into a struct.

Reply:
Well, that's a nice little lib, I've starred it.:) I'd usually say it would be nice to have it in the SDK too, but since it's that simple with the lib, we're not making it much simpler for the developers.

Reply:
I didn't use the above lib myself, but it looks like parsing is as simple as writing `mapstructure.Decode()`. With that, I don't think we can make it easier for the developers.

Reply:
After thinking about this, it would be a nice UX if I could do something like this:
```
type Config struct {
    GreaterThan2 int `validate:"required,gt=2"`
}

func (s *Source) Configure(ctx context.Context, raw map[string]string) error {
    var cfg Config
    err := sdk.ParseConfig(raw, &cfg)
    if err != nil {
        return err
    }
    // I have a validated and parsed cfg at this point
    // Do custom validations here if needed
    // ...
}
```

Having a function that parses and validates could make it simpler for our developers (_in addition to `sdk.ValidateConfig`_, which only does the validation part, in case people want to manually parse it).

Reply:
@lovromazgon agreed that it would be nice to have a function that would validate+parse.